### PR TITLE
Employee Qualification Service Refactoring

### PR DESCRIPTION
**Pull request by MathildaShongwe15**
image

### CODE REVIEW

File: HRIS.Services.Tests/Services/EmployeeQualificationServiceUnitTests.cs [hunk 1]
```diff
@@ -13,19 +13,42 @@ namespace HRIS.Services.Tests.Services;
 
 public class EmployeeQualificationServiceUnitTests
 {
-    private const int EmployeeId = 1;
-    private const int QualificationId = 1;
     private readonly Mock<IUnitOfWork> _db;
     private readonly IEmployeeQualificationService _employeeQualificationService;
     private readonly Mock<IEmployeeService> _employeeService;
+    private readonly Mock<AuthorizeIdentityMock> _identity;
+
```
Comment (EduanO) [inline HRIS.Services.Tests/Services/EmployeeQualificationServiceUnitTests.cs:20]:
```suggestion
```

File: HRIS.Services.Tests/Services/EmployeeQualificationServiceUnitTests.cs [hunk 2]
```diff
@@ -13,19 +13,42 @@ namespace HRIS.Services.Tests.Services;
...
 
     public EmployeeQualificationServiceUnitTests()
     {
         _db = new Mock<IUnitOfWork>();
+        _identity = new Mock<AuthorizeIdentityMock>();
         _employeeService = new Mock<IEmployeeService>();
-        _employeeQualificationService = new EmployeeQualificationService(_db.Object, _employeeService.Object);
+        _employeeQualificationService = new EmployeeQualificationService(_db.Object, _identity.Object);
+    }
```
Comment (EduanO) [inline HRIS.Services.Tests/Services/EmployeeQualificationServiceUnitTests.cs:27]:
```suggestion
    }
    
```

File: HRIS.Services.Tests/Services/EmployeeQualificationServiceUnitTests.cs [hunk 3]
```diff
@@ -126,138 +149,164 @@ public async Task GetEmployeeQualificationsByEmployeeId_Success()
         Assert.NotNull(result);
         Assert.Equivalent(EmployeeQualificationTestData.EmployeeQualification.ToDto(), result);
 
-        _employeeService.Verify(x => x.GetEmployeeById(It.IsAny<int>()), Times.Once);
         _db.Verify(
             x => x.EmployeeQualification.FirstOrDefault(It.IsAny<Expression<Func<EmployeeQualification, bool>>>()),
             Times.Once);
     }
 
+    #endregion
```
Comment (EduanO) [inline HRIS.Services.Tests/Services/EmployeeQualificationServiceUnitTests.cs:157]:
```suggestion
    #endregion
    
```

File: HRIS.Services.Tests/Services/EmployeeQualificationServiceUnitTests.cs [hunk 4]
```diff
@@ -126,138 +149,164 @@ public async Task GetEmployeeQualificationsByEmployeeId_Success()
...
-        _employeeService.Verify(x => x.GetEmployeeById(It.IsAny<int>()), Times.Once);
         _db.Verify(
             x => x.EmployeeQualification.FirstOrDefault(It.IsAny<Expression<Func<EmployeeQualification, bool>>>()),
             Times.Once);
         _db.Verify(x => x.EmployeeQualification.Update(It.IsAny<EmployeeQualification>()), Times.Once);
     }
 
     [Fact]
-    public async Task UpdateEmployeeQualification_Failure_GetEmployeeById()
+    public async Task UpdateEmployeeQaulification_DoesNotExist()
```
Comment (EduanO) [inline HRIS.Services.Tests/Services/EmployeeQualificationServiceUnitTests.cs]:
```suggestion
    public async Task UpdateEmployeeQualification_DoesNotExist()
```

File: HRIS.Services/Interfaces/IEmployeeQualificationService.cs [hunk 5]
```diff
@@ -4,14 +4,22 @@ namespace HRIS.Services.Interfaces;
 
 public interface IEmployeeQualificationService
 {
+
+
+    /// <summary>
+    ///     Check if user exist
+    /// </summary>
+    /// <param name="email"></param>
```
Comment (EduanO) [inline HRIS.Services/Interfaces/IEmployeeQualificationService.cs]:
```suggestion
    /// <param name="id"></param>
```

File: HRIS.Services/Interfaces/IEmployeeQualificationService.cs [hunk 6]
```diff
@@ -4,14 +4,22 @@ namespace HRIS.Services.Interfaces;
...
+    /// <returns>Boolean to check if the Termination exists</returns>
+    Task<bool> CheckIfExists(int Id);
```
Comment (EduanO) [inline HRIS.Services/Interfaces/IEmployeeQualificationService.cs]:
```suggestion
    Task<bool> CheckIfExists(int id);
```

File: HRIS.Services/Services/EmployeeQualificationService.cs [hunk 7]
```diff
@@ -9,25 +10,31 @@ namespace HRIS.Services.Services;
...
-    public EmployeeQualificationService(IUnitOfWork db, IEmployeeService employeeService)
+    private readonly AuthorizeIdentity _identity;
+    public EmployeeQualificationService(IUnitOfWork db, AuthorizeIdentity identity)
     {
         _db = db;
-        _employeeService = employeeService;
+        _identity = identity;
+    }
+
+    public async Task<bool> CheckIfExists(int Id)
```
Comment (EduanO) [inline HRIS.Services/Services/EmployeeQualificationService.cs]:
```suggestion
    public async Task<bool> CheckIfExists(int id)
```

**Review by EduanO — CHANGES_REQUESTED** [review]

File: HRIS.Services/Services/EmployeeQualificationService.cs [hunk 8]
```diff
@@ -41,20 +48,27 @@ public async Task<List<EmployeeQualificationDto>> GetAllEmployeeQualifications()
...
     public async Task<EmployeeQualificationDto> GetEmployeeQualificationsByEmployeeId(int employeeId)
     {
-        var employee = await _employeeService.GetEmployeeById(employeeId);
-        var qualifications = await _db.EmployeeQualification.FirstOrDefault(x => x.EmployeeId == employee.Id);
-        return qualifications?.ToDto();
+        if (_identity.IsSupport == false && _identity.EmployeeId != employeeId)
+            throw new CustomException("Unauthorized access");
+
+        var qualifications = await _db.EmployeeQualification.FirstOrDefault(x => x.EmployeeId == _identity.EmployeeId);
+        return qualifications.ToDto();
```
Comment (bseptember) [inline HRIS.Services/Services/EmployeeQualificationService.cs]:
```suggestion
     if (qualifications == null)
         throw new CustomException("No employee qualifications data.");

     return qualifications.ToDto();
```

File: HRIS.Services/Services/EmployeeQualificationService.cs [hunk 9]
```diff
@@ -41,20 +48,27 @@ public async Task<List<EmployeeQualificationDto>> GetAllEmployeeQualifications()
...
+        var exists = await CheckIfExists(employeeQualificationDto.Id);
+
+        if (!exists)
+            throw new CustomException("Employee Does Not Have Existing Qualifications Saved");
+
+        if (_identity.IsSupport == false && _identity.EmployeeId != employeeQualificationDto.EmployeeId)
+            throw new CustomException("Unauthorized access.");
 
         var qualification = await _db.EmployeeQualification.FirstOrDefault(x => x.Id == employeeQualificationDto.Id);
-        if (qualification == null) throw new CustomException("Employee Does Not Have Existing Qualifications Saved");
```
Comment (bseptember) [inline HRIS.Services/Services/EmployeeQualificationService.cs:73]:
```suggestion
      if (qualification == null)
          throw new CustomException("No employee qualification data.");
```

**Review by bseptember — CHANGES_REQUESTED** [review]
Please note, you'll need to add a null check unit test for those suggested changes.

**Review by EduanO — APPROVED** [review]

**Review by bseptember — APPROVED** [review]

**Review by ReitumetseMoloto — APPROVED** [review]

**Pull request merged**